packages/file_selector/file_selector_android/android/src/main/java/io/flutter/plugins/file_selector/PathUtils.java

```diff
@@ -0,0 +1,278 @@
+    return copyFilesToInternalStorage(uri, context, "");
+  }
+
+  private static String copyFileToInternalStorage(Uri uri, Context context, String cacheFolder) {
+    ArrayList<Uri> uris = new ArrayList<Uri>() {};
+    uris.add(uri);
+    ArrayList<String> file = copyFilesToInternalStorage(uris, context, cacheFolder);
+
+    return file.get(0);
+  }
+
+  @NonNull
+  public static ArrayList<String> copyFilesToInternalStorage(
+      ArrayList<Uri> uris, Context context, String cacheFolderName) {
+    ArrayList<String> absolutePaths = new ArrayList<>();
+
+    String newDirPath = context.getFilesDir() + "/" + cacheFolderName;
+    createDirectoryIfNotExists(newDirPath);
+
+    for (Uri uri : uris) {
+      String name =
+          getFileName(
+              uri, context, new String[] {OpenableColumns.DISPLAY_NAME, OpenableColumns.SIZE});
+      File output = getFile(context, newDirPath, uri, name);
+      absolutePaths.add(output.getAbsolutePath());
+    }
+
+    return absolutePaths;
+  }
+
+  private static File getFile(Context context, String newDirPath, Uri uri, String name) {
+    File output;
+    output = new File(newDirPath + "/" + name);
+    try {
+      InputStream inputStream = context.getContentResolver().openInputStream(uri);
+      FileOutputStream outputStream = new FileOutputStream(output);
+      int read;
+      int bufferSize = 1024;
+      final byte[] buffers = new byte[bufferSize];
+      while ((read = inputStream.read(buffers)) != -1) {
+        outputStream.write(buffers, 0, read);
+      }
+
+      inputStream.close();
+      outputStream.close();
+
+    } catch (Exception e) {
+      System.out.println("There was an error adding a file to the application cache");
+    }
+    return output;
+  }
+
+  private static void createDirectoryIfNotExists(String newDirPath) {
+    File dir = new File(newDirPath);
+    if (!dir.exists()) {
+      dir.mkdir();
+    }
+  }
+
+  /**
+   * Returns an absolute path for the file to be saved from the 3 possible content paths to return.
+   * If none of these match, then we will create a copy of the created file in the internal storage
+   * and return that path.
+   */
+  @VisibleForTesting
+  @Nullable
+  public static String getSavePathUri(final Uri uri, Context context) {
+    final String docId = DocumentsContract.getDocumentId(uri);
+
+    if (docId.isEmpty()) {
+      return null;
+    }
+
+    switch (uri.getAuthority()) {
+      case externalStorageDocuments:
+        final String[] split = docId.split(":");
+        return getPathFromExtSD(uri, context, split);
+      case providersDownloadsDocuments:
+        return getDownloadsDocumentPath(uri, context, docId);
+      case providersMediaDocuments:
+        return getMediaDocumentDataColumn(context, docId);
+    }
+
+    if (uri.getScheme().equalsIgnoreCase("content")) {
+      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
+        return copyFileToInternalStorage(uri, context, cacheFolder);
```

Comment:
Why would selecting a save location involve copying files? If this is actually necessary, it definitely need an explanatory comment.

(Also, if for some reason we need to copy files, it shouldn't be done on the platform thread.)

Reply:
Addressed in the other comment. [Issue created](https://github.com/flutter/flutter/issues/113441) and save path not included in the first version.